tfx_addons/sampler/executor.py

```diff
@@ -0,0 +1,259 @@
+    splits = json_utils.loads(exec_properties[spec.SAMPLER_SPLIT_KEY])
+    copy_others = exec_properties[spec.SAMPLER_COPY_KEY]
+    shards = exec_properties[spec.SAMPLER_SHARDS_KEY]
+    null_classes = json_utils.loads(exec_properties[spec.SAMPLER_CLASSES_KEY])
+
+    input_artifact = artifact_utils.get_single_instance(
+        input_dict[spec.SAMPLER_INPUT_KEY])
+    output_artifact = artifact_utils.get_single_instance(
+        output_dict[spec.SAMPLER_OUTPUT_KEY])
+
+    if copy_others:
+      output_artifact.split_names = input_artifact.split_names
+    else:
+      output_artifact.split_names = artifact_utils.encode_split_names(splits)
+
+    # Fetch the input uri for each split
+    split_data = {}
+    for split in artifact_utils.decode_split_names(input_artifact.split_names):
+      uri = artifact_utils.get_split_uri([input_artifact], split)
+      split_data[split] = uri
+
+    for split, uri in split_data.items():
+      if split in splits:  # Undersampling split
+        output_dir = artifact_utils.get_split_uri([output_artifact], split)
+        split_dir = os.path.join(output_dir, f"Split-{split}")
+        with self._CreatePipeline(split_dir) as p:
+          sample(p, uri, label, shards, null_classes, split_dir,
+                 sampling_strategy)
+      elif copy_others:  # Copy the other split if copy_others is True
+        input_dir = uri
+        output_dir = artifact_utils.get_split_uri([output_artifact], split)
+        for filename in fileio.listdir(input_dir):
+          input_uri = os.path.join(input_dir, filename)
+          output_uri = os.path.join(output_dir, filename)
+          io_utils.copy_file(src=input_uri, dst=output_uri, overwrite=True)
+
+
+def generate_elements(x, label):
+  """Funciton that fetches the class label from a tf.Example and returns one
+  item in a K-V PCollection with the key as the label and the value as the
+  string-parsed tf.Example."""
+
+  class_label = None
+  parsed = tf.train.Example.FromString(x.numpy())
+  if parsed.features.feature[label].int64_list.value:
+    val = parsed.features.feature[label].int64_list.value
+    if len(val) > 0:
+      class_label = val[0]
+  else:
+    val = parsed.features.feature[label].bytes_list.value
+    if len(val) > 0:
+      class_label = val[0].decode()
+  return (class_label, parsed)
+
+
+def sample_data(_,
+                val,
+                sampling_strategy=spec.SamplingStrategy.UNDERSAMPLE,
+                side=0):
+  """Function called in a Beam pipeline that performs sampling using Python's
+  random module on an input key:value pair, where the key is the class label
+  and the values are the data points to sample. Note that the key is discarded."""
+
+  if sampling_strategy == spec.SamplingStrategy.UNDERSAMPLE:
+    random_sample_data = random.sample(val, side)
+  elif sampling_strategy == spec.SamplingStrategy.OVERSAMPLE:
+    random_sample_data = random.choices(val, k=side)
```

Comment:
This implicitly assumes the enum `SamplingStrategy` only has 2 values. Consider an `else` case raising a ValueError to make that assumption explicit rather than implicit